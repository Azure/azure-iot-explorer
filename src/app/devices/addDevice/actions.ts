import actionCreatorFactory from 'typescript-fsa';
import { ADD } from '../../constants/actionTypes';
import { DEVICE } from '../../constants/actionPrefixes';
import { DeviceIdentity } from '../../api/models/deviceIdentity';

const deviceListCreator = actionCreatorFactory(DEVICE);
export const addDeviceAction = deviceListCreator.async<DeviceIdentity, DeviceIdentity>(ADD);
