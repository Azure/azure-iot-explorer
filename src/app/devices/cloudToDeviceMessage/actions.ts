/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import actionCreatorFactory from 'typescript-fsa';
import { DEVICECONTENT } from '../../constants/actionPrefixes';
import { CLOUD_TO_DEVICE_MESSAGE } from '../../constants/actionTypes';

export const cloudToDeviceMessageAction = actionCreatorFactory(DEVICECONTENT).async<CloudToDeviceMessageActionParameters, string>(CLOUD_TO_DEVICE_MESSAGE);

export interface CloudToDeviceMessageActionParameters {
    deviceId: string;
    body: string;
    properties?: Array<{key: string, value: string, isSystemProperty: boolean}>;
}
