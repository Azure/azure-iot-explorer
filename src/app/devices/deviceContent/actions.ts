/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import actionCreatorFactory from 'typescript-fsa';
import * as actionPrefixes from '../../constants/actionPrefixes';
import * as actionTypes from '../../constants/actionTypes';
import { ModelDefinition } from '../../api/models/modelDefinition';
import { Twin } from '../../api/models/device';
import { DeviceIdentity } from '../../api/models/deviceIdentity';
import { DigitalTwinInterfaces } from './../../api/models/digitalTwinModels';
import { REPOSITORY_LOCATION_TYPE } from './../../constants/repositoryLocationTypes';

const deviceContentCreator = actionCreatorFactory(actionPrefixes.DEVICECONTENT);
const clearModelDefinitionsAction = deviceContentCreator(actionTypes.CLEAR_MODEL_DEFINITIONS);
const cloudToDeviceMessageAction = deviceContentCreator.async<CloudToDeviceMessageActionParameters, string>(actionTypes.CLOUD_TO_DEVICE_MESSAGE);
const getDeviceIdentityAction = deviceContentCreator.async<string, DeviceIdentity> (actionTypes.GET_DEVICE_IDENTITY);
const getDigitalTwinInterfacePropertiesAction = deviceContentCreator.async<string, DigitalTwinInterfaces>(actionTypes.GET_DIGITAL_TWIN_INTERFACE_PROPERTIES);
const getTwinAction = deviceContentCreator.async<string, Twin>(actionTypes.GET_TWIN);
const getModelDefinitionAction = deviceContentCreator.async<GetModelDefinitionActionParameters, ModelDefinitionActionResult>(actionTypes.FETCH_MODEL_DEFINITION);
const invokeDirectMethodAction = deviceContentCreator.async<InvokeMethodActionParameters, string>(actionTypes.INVOKE_DEVICE_METHOD);
const invokeDigitalTwinInterfaceCommandAction = deviceContentCreator.async<InvokeDigitalTwinInterfaceCommandActionParameters, string>(actionTypes.INVOKE_DIGITAL_TWIN_INTERFACE_COMMAND);
const patchDigitalTwinInterfacePropertiesAction = deviceContentCreator.async<PatchDigitalTwinInterfacePropertiesActionParameters, DigitalTwinInterfaces>(actionTypes.PATCH_DIGITAL_TWIN_INTERFACE_PROPERTIES);
const setInterfaceIdAction = deviceContentCreator<string>(actionTypes.SET_INTERFACE_ID);
const updateDeviceIdentityAction = deviceContentCreator.async<DeviceIdentity, DeviceIdentity> (actionTypes.UPDATE_DEVICE_IDENTITY);
const updateTwinAction = deviceContentCreator.async<UpdateTwinActionParameters, Twin>(actionTypes.UPDATE_TWIN);
export {
    clearModelDefinitionsAction,
    cloudToDeviceMessageAction,
    getDeviceIdentityAction,
    getDigitalTwinInterfacePropertiesAction,
    getTwinAction,
    getModelDefinitionAction,
    invokeDirectMethodAction,
    invokeDigitalTwinInterfaceCommandAction,
    patchDigitalTwinInterfacePropertiesAction,
    setInterfaceIdAction,
    updateTwinAction,
    updateDeviceIdentityAction
};

export interface CloudToDeviceMessageActionParameters {
    deviceId: string;
    body: string;
    properties?: Array<{key: string, value: string, isSystemProperty: boolean}>;
}

export interface PatchDigitalTwinInterfacePropertiesActionParameters {
    digitalTwinId: string;
    interfacesPatchData: any; // tslint:disable-line:no-any
    propertyKey: string;
}

export interface InvokeDigitalTwinInterfaceCommandActionParameters {
    digitalTwinId: string;
    commandName: string;
    commandPayload: any; // tslint:disable-line:no-any
    propertyKey?: string;
}

export interface InvokeMethodActionParameters {
    connectTimeoutInSeconds: number;
    deviceId: string;
    methodName: string;
    payload?: object;
    responseTimeoutInSeconds: number;
}

export interface UpdateTwinActionParameters {
    deviceId: string;
    twin: Twin;
}

export interface ModelDefinitionActionResult {
    modelDefinition: ModelDefinition;
    source: REPOSITORY_LOCATION_TYPE;
}

export interface GetModelDefinitionActionParameters {
    digitalTwinId: string;
    interfaceId: string;
}
