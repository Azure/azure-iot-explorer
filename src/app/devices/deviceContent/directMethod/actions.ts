/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import actionCreatorFactory from 'typescript-fsa';
import * as actionPrefixes from '../../../constants/actionPrefixes';
import * as actionTypes from '../../../constants/actionTypes';

export const invokeDirectMethodAction = actionCreatorFactory(actionPrefixes.DEVICECONTENT).async<InvokeMethodActionParameters, string>(actionTypes.INVOKE_DEVICE_METHOD);

export interface InvokeMethodActionParameters {
    connectTimeoutInSeconds: number;
    deviceId: string;
    methodName: string;
    payload?: object;
    responseTimeoutInSeconds: number;
}
