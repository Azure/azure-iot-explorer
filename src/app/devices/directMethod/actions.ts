/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import actionCreatorFactory from 'typescript-fsa';
import { DEVICECONTENT } from '../../constants/actionPrefixes';
import { INVOKE_DEVICE_METHOD } from '../../constants/actionTypes';

export const invokeDirectMethodAction = actionCreatorFactory(DEVICECONTENT).async<InvokeMethodActionParameters, string>(INVOKE_DEVICE_METHOD);

export interface InvokeMethodActionParameters {
    connectTimeoutInSeconds: number;
    deviceId: string;
    methodName: string;
    payload?: string;
    responseTimeoutInSeconds: number;
}
