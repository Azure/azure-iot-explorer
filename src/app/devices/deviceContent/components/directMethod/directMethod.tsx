/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { RouteComponentProps } from 'react-router-dom';
import { TextField } from 'office-ui-fabric-react/lib/TextField';
import { CommandBar } from 'office-ui-fabric-react/lib/CommandBar';
import { Slider, SliderBase } from 'office-ui-fabric-react/lib/Slider';
import { Spinner, SpinnerSize } from 'office-ui-fabric-react/lib/Spinner';
import { LocalizationContextConsumer, LocalizationContextInterface } from '../../../../shared/contexts/localizationContext';
import { ResourceKeys } from '../../../../../localization/resourceKeys';
import { InvokeMethodActionParameters } from '../../actions';
import { getDeviceIdFromQueryString } from '../../../../shared/utils/queryStringHelper';
import { DIRECT_METHOD } from '../../../../constants/iconNames';
import LabelWithTooltip from '../../../../shared/components/labelWithTooltip';
import { ThemeContextConsumer, ThemeContextInterface } from '../../../../shared/contexts/themeContext';
const EditorPromise = import('react-monaco-editor');
const Editor = React.lazy(() => EditorPromise);
import '../../../../css/_deviceDetail.scss';

const SLIDER_MAX = 300;

export interface DirectMethodState {
    connectionTimeout: number;
    methodName: string;
    responseTimeout: number;
    payload: string;
}

export interface DirectMethodProps {
    onInvokeMethodClick: (properties: InvokeMethodActionParameters) => void;
}

export default class DirectMethod extends React.Component<DirectMethodProps & RouteComponentProps, DirectMethodState> {

    private connectionSliderRef: React.Ref<SliderBase>;

    constructor(props: DirectMethodProps & RouteComponentProps) {
        super(props);
        this.state = {
            connectionTimeout: 0,
            methodName: '',
            payload: '',
            responseTimeout: 0
        };
    }

    public render(): JSX.Element {
        return (
            <LocalizationContextConsumer>
                {(context: LocalizationContextInterface) => (
                    <>
                        {this.showCommandBar(context)}
                        <h3>{context.t(ResourceKeys.directMethod.headerText)}</h3>
                        <div className="device-detail">
                            {this.renderMethodsName(context)}
                            {this.renderMethodsPayloadSection(context)}
                        </div>
                    </>
                )}
            </LocalizationContextConsumer>
        );
    }

    private readonly showCommandBar = (context: LocalizationContextInterface) => {
        return (
            <CommandBar
                className="command"
                items={[
                    {
                        ariaLabel: context.t(ResourceKeys.directMethod.invokeMethodButtonText),
                        disabled: !this.formReady(),
                        iconProps: {iconName: DIRECT_METHOD},
                        key: DIRECT_METHOD,
                        name: context.t(ResourceKeys.directMethod.invokeMethodButtonText),
                        onClick: this.onInvokeMethodClick
                    }
                ]}
            />
        );
    }

    private readonly formReady = (): boolean => {
        const { methodName, payload } = this.state;
        return !!methodName && (!payload || this.isValidJson(payload));
    }

    private isValidJson = (content: string) => {
        try {
            JSON.parse(content);
            return true;
        } catch {
            return false;
        }
    }

    private readonly onInvokeMethodClick = () => {
        const { connectionTimeout, methodName, responseTimeout, payload } = this.state;

        this.props.onInvokeMethodClick({
            connectTimeoutInSeconds: connectionTimeout,
            deviceId: getDeviceIdFromQueryString(this.props),
            methodName,
            payload: payload ? JSON.parse(payload) : {},
            responseTimeoutInSeconds: responseTimeout
        });
    }

    private readonly renderMethodsName = (context: LocalizationContextInterface) => {
        return (
                <TextField
                    label={context.t(ResourceKeys.directMethod.methodName)}
                    ariaLabel={context.t(ResourceKeys.directMethod.methodName)}
                    value={this.state.methodName}
                    onChange={this.onMethodNameChange}
                    required={true}
                    placeholder={context.t(ResourceKeys.directMethod.methodNamePlaceHolder)}
                />
        );
    }

    private readonly renderMethodsPayloadSection = (context: LocalizationContextInterface) => {
        const { connectionTimeout, responseTimeout, payload } = this.state;

        return (
            <div className="method-payload">
                <LabelWithTooltip
                    tooltipText={context.t(ResourceKeys.directMethod.payloadTooltip)}
                >
                    {context.t(ResourceKeys.directMethod.payload)}
                </LabelWithTooltip>
                <div className="direct-method-monaco-editor">
                    <React.Suspense fallback={<Spinner title={'loading'} size={SpinnerSize.large} />}>
                        <ThemeContextConsumer>
                            {(themeContext: ThemeContextInterface) => (
                                <Editor
                                    language="json"
                                    options={{
                                        automaticLayout: true
                                    }}
                                    height="25vh"
                                    value={payload}
                                    onChange={this.onEditorChange}
                                    theme={themeContext.monacoTheme}
                                />
                            )}
                        </ThemeContextConsumer>
                    </React.Suspense>
                </div>
                <LabelWithTooltip
                    tooltipText={context.t(ResourceKeys.directMethod.connectionTimeoutTooltip)}
                >
                    {context.t(ResourceKeys.directMethod.connectionTimeout)}
                </LabelWithTooltip>
                <Slider
                    ariaLabel={context.t(ResourceKeys.directMethod.connectionTimeout)}
                    min={0}
                    max={SLIDER_MAX}
                    ref={this.connectionSliderRef}
                    value={connectionTimeout}
                    onChange={this.onConnectionTimeoutChange}
                />
                <LabelWithTooltip
                    tooltipText={context.t(ResourceKeys.directMethod.responseTimeoutTooltip)}
                >
                    {context.t(ResourceKeys.directMethod.responseTimeout)}
                </LabelWithTooltip>
                <Slider
                    ariaLabel={context.t(ResourceKeys.directMethod.responseTimeout)}
                    min={connectionTimeout}
                    max={SLIDER_MAX}
                    value={responseTimeout}
                    onChange={this.onMethodTimeoutChange}
                />
            </div>
        );
    }

    private readonly onEditorChange = (value: string) => {
        this.setState({
            payload: value
        });
    }

    private readonly onConnectionTimeoutChange = (value: number) => {
        this.setState({
            connectionTimeout: value,
            responseTimeout: Math.max(this.state.responseTimeout, value)
        });
    }

    private readonly onMethodTimeoutChange = (value: number) => {
        this.setState({
            responseTimeout: value
        });
    }

    private readonly onMethodNameChange = (event: React.FormEvent, value?: string) => {
        if (!!value) {
            this.setState({
                methodName: value
            });
        }
    }
}
