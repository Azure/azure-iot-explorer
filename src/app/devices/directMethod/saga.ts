/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { call, put, takeEvery } from 'redux-saga/effects';
import { SagaIterator } from 'redux-saga';
import { Action } from 'typescript-fsa';
import { invokeDirectMethodAction, InvokeMethodActionParameters } from './actions';
import { invokeDirectMethod } from '../../api/services/devicesService';
import { raiseNotificationToast } from '../../notifications/components/notificationToast';
import { NotificationType } from '../../api/models/notification';
import { ResourceKeys } from '../../../localization/resourceKeys';
import { InvokeMethodParameters } from '../../api/parameters/deviceParameters';

export function* invokeDirectMethodSagaWorker(action: Action<InvokeMethodActionParameters>): SagaIterator {
    const toastId: number = Math.random();

    try {
        const invokeMethodParameters: InvokeMethodParameters = {
            ...action.payload,
        };

        yield call(notifyMethodInvokedHelper, toastId, invokeMethodParameters);

        const response = yield call(invokeDirectMethod, invokeMethodParameters);
        const stringifiedResponse = typeof response === 'object' ? JSON.stringify(response) : response;

        yield call(raiseNotificationToast, {
            id: toastId,
            text: {
                translationKey: ResourceKeys.notifications.invokeMethodOnSuccess,
                translationOptions: {
                    deviceId: action.payload.deviceId,
                    methodName: action.payload.methodName,
                    response: stringifiedResponse
                },
            },
            type: NotificationType.success
        });

        yield put(invokeDirectMethodAction.done({params: action.payload, result: stringifiedResponse}));
    } catch (error) {
        yield call(raiseNotificationToast, {
            id: toastId,
            text: {
                translationKey: ResourceKeys.notifications.invokeMethodOnError,
                translationOptions: {
                    deviceId: action.payload.deviceId,
                    error,
                },
            },
            type: NotificationType.error
        });

        yield put(invokeDirectMethodAction.failed({params: action.payload, error}));
    }
}

export function* notifyMethodInvokedHelper(toastId: number, payload: InvokeMethodParameters) {
    if (payload) {
        yield call(raiseNotificationToast, {
            id: toastId,
            text: {
                translationKey: ResourceKeys.notifications.invokingMethodWithPayload,
                translationOptions: {
                    deviceId: payload.deviceId,
                    methodName: payload.methodName,
                    payload: JSON.stringify(payload.payload),
                },
            },
            type: NotificationType.info
        });
        return payload.payload;
    }
}

export function* invokeDirectMethodSaga() {
    yield takeEvery(invokeDirectMethodAction.started.type, invokeDirectMethodSagaWorker);
}
