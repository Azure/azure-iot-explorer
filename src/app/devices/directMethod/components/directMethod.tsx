/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { useLocation } from 'react-router-dom';
import { CommandBar } from '@fluentui/react';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import { getDeviceIdFromQueryString } from '../../../shared/utils/queryStringHelper';
import { DIRECT_METHOD } from '../../../constants/iconNames';
import { HeaderView } from '../../../shared/components/headerView';
import { useAsyncSagaReducer } from '../../../shared/hooks/useAsyncSagaReducer';
import { invokeDirectMethodSaga } from '../saga';
import { invokeDirectMethodAction } from '../actions';
import { DirectMethodForm } from './directMethodForm';
import '../../../css/_deviceDetail.scss';
import { AppInsightsClient } from '../../../shared/appTelemetry/appInsightsClient';
import { TELEMETRY_PAGE_NAMES, TELEMETRY_USER_ACTIONS } from '../../../../app/constants/telemetry';

const DEFAULT_TIMEOUT = 10;

export const DirectMethod: React.FC = () => {
    const { t } = useTranslation();
    const { search } = useLocation();
    const deviceId = getDeviceIdFromQueryString(search);

    const [ , dispatch ] = useAsyncSagaReducer(() => undefined, invokeDirectMethodSaga, undefined);
    const [connectionTimeOut, setConnectionTimeOut] = React.useState<number>(DEFAULT_TIMEOUT);
    const [methodName, setMethodName] = React.useState<string>('');
    const [payload, setPayload] = React.useState<string>('');
    const [responseTimeOut, setResponseTimeOut] = React.useState<number>(DEFAULT_TIMEOUT);

    React.useEffect(() => {
        AppInsightsClient.getInstance()?.trackPageView({name: TELEMETRY_PAGE_NAMES.DEVICE_DIRECT_METHOD});
    }, []); // tslint:disable-line: align

    const showCommandBar = () => {
        return (
            <CommandBar
                className="command"
                items={[
                    {
                        ariaLabel: t(ResourceKeys.directMethod.invokeMethodButtonText),
                        disabled: !formReady(),
                        iconProps: {iconName: DIRECT_METHOD},
                        key: DIRECT_METHOD,
                        name: t(ResourceKeys.directMethod.invokeMethodButtonText),
                        onClick: onInvokeMethodClickHandler
                    }
                ]}
            />
        );
    };

    const formReady = (): boolean => !!methodName;
    const getPayload = (content: string) => {
        // payload could be json or simply string
        try {
            return JSON.parse(content);
        } catch {
            return content;
        }
    };

    const onInvokeMethodClickHandler = () => {
        AppInsightsClient.trackUserAction(TELEMETRY_USER_ACTIONS.DEVICE_INVOKE_DIRECT_METHOD);
        dispatch(invokeDirectMethodAction.started({
            connectTimeoutInSeconds: connectionTimeOut,
            deviceId,
            methodName,
            payload: getPayload(payload),
            responseTimeoutInSeconds: responseTimeOut
        }));
    };

    return (
        <>
            {showCommandBar()}
            <HeaderView
                headerText={ResourceKeys.directMethod.headerText}
                tooltip={ResourceKeys.directMethod.tooltip}
            />
            <DirectMethodForm
                methodName={methodName}
                connectionTimeOut={connectionTimeOut}
                responseTimeOut={responseTimeOut}
                setMethodName={setMethodName}
                setPayload={setPayload}
                setConnectionTimeOut={setConnectionTimeOut}
                setResponseTimeOut={setResponseTimeOut}
            />
        </>
    );
};
