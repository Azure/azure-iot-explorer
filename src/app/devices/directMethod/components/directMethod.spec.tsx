/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import 'jest';
import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { CommandBar } from '@fluentui/react';
import { DirectMethod } from './directMethod';
import { invokeDirectMethodAction } from '../actions';

jest.mock('react-router-dom', () => ({
    useLocation: () => ({ search: '?deviceId=test' })
}));

describe('directMethod', () => {
    it('matches snapshot', () => {
        expect(shallow(<DirectMethod/>)).toMatchSnapshot();
    });

    it('invokes direct method', () => {
        const mockGetDeviceTwin = jest.spyOn(invokeDirectMethodAction, 'started');
        const realUseState = React.useState;
        // tslint:disable-next-line: no-magic-numbers
        jest.spyOn(React, 'useState').mockImplementationOnce(() => realUseState(60));
        jest.spyOn(React, 'useState').mockImplementationOnce(() => realUseState('method'));
        jest.spyOn(React, 'useState').mockImplementationOnce(() => realUseState(JSON.stringify({foo: 'bar'})));
        // tslint:disable-next-line: no-magic-numbers
        jest.spyOn(React, 'useState').mockImplementationOnce(() => realUseState(20));

        const wrapper = mount(<DirectMethod/>);
        const commandBar = wrapper.find(CommandBar).first();
        act(() => commandBar.props().items[0].onClick(null));
        expect(mockGetDeviceTwin.mock.calls[0][0]).toEqual({
            connectTimeoutInSeconds: 60,
            deviceId: 'test',
            methodName: 'method',
            payload: {foo: 'bar'},
            responseTimeoutInSeconds: 20
        });
    });
});
