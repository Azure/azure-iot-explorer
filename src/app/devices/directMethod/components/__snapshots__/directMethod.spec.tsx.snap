// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`directMethod matches snapshot 1`] = `
<Fragment>
  <StyledCommandBarBase
    className="command"
    items={
      Array [
        Object {
          "ariaLabel": "directMethod.invokeMethodButtonText",
          "disabled": true,
          "iconProps": Object {
            "iconName": "Remote",
          },
          "key": "Remote",
          "name": "directMethod.invokeMethodButtonText",
          "onClick": [Function],
        },
      ]
    }
  />
  <Component
    headerText="directMethod.headerText"
    tooltip="directMethod.tooltip"
  />
  <div
    className="device-detail"
  >
    <StyledTextFieldBase
      ariaLabel="directMethod.methodName"
      label="directMethod.methodName"
      onChange={[Function]}
      placeholder="directMethod.methodNamePlaceHolder"
      required={true}
      value=""
    />
    <div
      className="method-payload"
    >
      <Component
        tooltipText="directMethod.payloadTooltip"
      >
        directMethod.payload
      </Component>
      <StyledTextFieldBase
        className="payload-input"
        multiline={true}
        onChange={[Function]}
        rows={5}
      />
      <Component
        tooltipText="directMethod.connectionTimeoutTooltip"
      >
        directMethod.connectionTimeout
      </Component>
      <StyledSliderBase
        ariaLabel="directMethod.connectionTimeout"
        max={300}
        min={0}
        onChange={[Function]}
        value={10}
      />
      <Component
        tooltipText="directMethod.responseTimeoutTooltip"
      >
        directMethod.responseTimeout
      </Component>
      <StyledSliderBase
        ariaLabel="directMethod.responseTimeout"
        max={300}
        min={10}
        onChange={[Function]}
        value={10}
      />
    </div>
  </div>
</Fragment>
`;
